packages/ui/components/overlays/src/OverlayController.js

```diff
@@ -1148,7 +1145,20 @@ export class OverlayController extends EventTarget {
 
   /** @private */
   __escKeyHandler(/** @type {KeyboardEvent} */ ev) {
-    return ev.key === 'Escape' && this.hide();
+    // @ts-expect-error
+    if (ev.key !== 'Escape' || ev._hasAlreadyHandledEscape) return;
+
+    this.hide();
+
+    const shouldNotCloseParents = this.hidesOnEsc && !this.hidesOnOutsideEsc;
```

Comment:
If the child overlay has `hidesOnOutsideEsc` set to true, does not mean the parent overlay should close.

Reply:
Good catch. Fixed it (and added tests for all possible oombinations, including shadow dom)